modules/host/network.go

```diff
@@ -49,6 +49,75 @@ func (h *Host) threadedUpdateHostname(closeChan chan struct{}) {
 	}
 }
 
+// threadedTrackWorkingState periodically checks if the host is working,
+// where working is defined as having received 3 settings calls in the past 15
+// minutes.
+func (h *Host) threadedTrackWorkingState(closeChan chan struct{}) {
+	defer close(closeChan)
+
+	for {
+		prevSettingsCalls := atomic.LoadUint64(&h.atomicSettingsCalls)
+
+		select {
+		case <-h.tg.StopChan():
+			return
+		case <-time.After(workingStateFrequency):
+		}
+
+		settingsCalls := atomic.LoadUint64(&h.atomicSettingsCalls)
+
+		// sanity check
+		if prevSettingsCalls > settingsCalls {
+			build.Severe("the host's settings calls decremented")
+			continue
+		}
+
+		h.mu.Lock()
+		if settingsCalls-prevSettingsCalls >= workingStateThreshold {
+			h.workingState = WorkingStateWorking
+		} else {
+			h.workingState = WorkingStateNotWorking
+		}
+		h.mu.Unlock()
+	}
+}
+
+// threadedTrackConnectabilityState periodically checks if the host is
+// connectable at its netaddress.
+func (h *Host) threadedTrackConnectabilityState(closeChan chan struct{}) {
+	defer close(closeChan)
+
+	for {
+		select {
+		case <-h.tg.StopChan():
+			return
+		case <-time.After(connectabilityCheckFrequency):
+		}
+
+		h.mu.RLock()
+		autoAddr := h.autoAddress
+		userAddr := h.settings.NetAddress
+		h.mu.RUnlock()
+
+		activeAddr := autoAddr
+		if userAddr != "" {
+			activeAddr = userAddr
+		}
+
+		conn, err := net.Dial("tcp", string(activeAddr))
```

Comment:
Need to use the quick-shutdown pattern on dial calls.

https://github.com/NebulousLabs/Sia/blob/master/modules/host/network.go#L141

^ example. It's pretty gross, but we've not found a better way to do it that's equally complete. Anything else we've figured out so far has been just as gross.